tsdb/store.go

```diff
@@ -678,6 +678,9 @@ func (s *Store) Shard(id uint64) *Shard {
 // longer in a "bad" state. This method will return any shards that
 // were removed from the cache.
 func (s *Store) ClearBadShardList() map[uint64]error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
I think this will deadlock on the mutex in `GetBadShardList`.  Please verify that.

Move it down below that call to just above the `clear`